File: evaluation/typescript/src/segmentEvaluator.ts
import { Segment, SegmentUser } from './proto/feature/segment_pb';
import { User } from './proto/user/user_pb';
import { RuleEvaluator } from './ruleEvaluator';
//
class SegmentEvaluator {
  // evaluate reports whether the user belongs to ANY of the given segments (OR).
  evaluate(
    segmentIDs: string[],
    user: User,
    segments: Map<string, Segment> | null,
    segmentUsers: SegmentUser[],
  ): boolean {
    for (const segmentID of segmentIDs) {
      const segment = segments?.get(segmentID) ?? null;
      if (this.isUserInSegment(segmentID, user, segment, segmentUsers)) {
        return true;
      }
    }
    return false;
  }

  // isUserInSegment reports whether the user belongs to the segment:
  // the user is in the included-user list OR matches any of the segment rules.
  private isUserInSegment(
    segmentID: string,
    user: User,
    segment: Segment | null,
    segmentUsers: SegmentUser[],
  ): boolean {
    // 1. Explicit include list — existing behavior, unchanged.
    if (
      this.containsSegmentUser(segmentID, user.getId(), SegmentUser.State.INCLUDED, segmentUsers)
    ) {
      return true;
    }
    // 2. Rules. Segment rules use the same Rule message as feature flag rules,
    // so the flag-side rule evaluator (OR across rules, AND across clauses,
    // attribute resolution, all operators) runs on them as is.
    // segmentUsers/segments/flagVariations are empty/null: validation rejects
    // SEGMENT and FEATURE_FLAG operators inside segment rules, and null makes
    // any such clause fail closed.
    if (!segment || segment.getRulesList().length === 0) {
      return false;
    }
    const ruleEvaluator = new RuleEvaluator();
    const matchedRule = ruleEvaluator.evaluate(segment.getRulesList(), user, [], null, null);
    return matchedRule !== null;
  }

  private containsSegmentUser(
    segmentID: string,
    userID: string,
    state: SegmentUser.StateMap[keyof SegmentUser.StateMap],
    segmentUsers: SegmentUser[],
  ): boolean {
    for (const user of segmentUsers) {
      if (user.getSegmentId() !== segmentID) {
        continue;
      }
      if (user.getUserId() !== userID) {
        continue;
      }
      if (user.getState() !== state) {
        continue;
      }
      return true;
    }
    return false;
  }
}

export { SegmentEvaluator };

File: evaluation/go/segment_evaluator.go
// Copyright 2026 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package evaluation

import (
	featureproto "github.com/bucketeer-io/bucketeer/v2/proto/feature"
	userproto "github.com/bucketeer-io/bucketeer/v2/proto/user"
)

type segmentEvaluator struct {
}

// Evaluate reports whether the user belongs to ANY of the given segments (OR).
func (e *segmentEvaluator) Evaluate(
	segmentIDs []string,
	user *userproto.User,
	segments map[string]*featureproto.Segment,
	segmentUsers []*featureproto.SegmentUser,
) (bool, error) {
	for _, segmentID := range segmentIDs {
		inSegment, err := e.isUserInSegment(segmentID, user, segments[segmentID], segmentUsers)
		if err != nil {
			return false, err
		}
		if inSegment {
			return true, nil
		}
	}
	return false, nil
}

// isUserInSegment reports whether the user belongs to the segment:
// the user is in the included-user list OR matches any of the segment rules.
func (e *segmentEvaluator) isUserInSegment(
	segmentID string,
	user *userproto.User,
	segment *featureproto.Segment,
	segmentUsers []*featureproto.SegmentUser,
) (bool, error) {
	// 1. Explicit include list — existing behavior, unchanged.
	if e.containsSegmentUser(segmentID, user.Id, featureproto.SegmentUser_INCLUDED, segmentUsers) {
		return true, nil
	}
	// 2. Rules. Segment.Rules is []*featureproto.Rule — the same type as Feature.Rules —
	// so the flag-side rule evaluator (OR across rules, AND across clauses, attribute
	// resolution, all operators) runs on it as is.
	// segmentUsers/segments/flagVariations are nil: validation rejects SEGMENT and
	// FEATURE_FLAG operators inside segment rules, and nil makes any such clause
	// fail closed.
	if segment == nil || len(segment.Rules) == 0 {
		return false, nil
	}
	ruleEval := &ruleEvaluator{} // local value: avoids the embedding cycle
	matchedRule, err := ruleEval.Evaluate(segment.Rules, user, nil, nil, nil)
	if err != nil {
		return false, err
	}
	return matchedRule != nil, nil
}

func (e *segmentEvaluator) containsSegmentUser(
	segmentID, userID string,
	state featureproto.SegmentUser_State,
	segmentUsers []*featureproto.SegmentUser,
) bool {
	for _, user := range segmentUsers {
		if user.SegmentId != segmentID {
			continue
		}
		if user.UserId != userID {
			continue
		}
		if user.State != state {
			continue
		}
		return true
	}
	return false
}
